### ISSUE TITLE
Max Transaction and Crash

### ISSUE THREAD

**Issue by andrewkkchan** [objectbox/objectbox-java#158]
Fatal Exception: java.lang.RuntimeException: Unable to create application com.mm.main.app.application.MyApplication: io.objectbox.exception.DbException: Maximum of threads/transactions reached
       at android.app.ActivityThread.handleBindApplication(ActivityThread.java:4867)
       at android.app.ActivityThread.access$1500(ActivityThread.java:178)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1531)
       at android.os.Handler.dispatchMessage(Handler.java:111)
       at android.os.Looper.loop(Looper.java:194)
       at android.app.ActivityThread.main(ActivityThread.java:5643)
       at java.lang.reflect.Method.invoke(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:372)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:960)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:755)
Caused by io.objectbox.exception.DbException: Maximum of threads/transactions reached
       at io.objectbox.BoxStore.nativeCreate(BoxStore.java)
       at io.objectbox.BoxStore.<init>(BoxStore.java:149)
       at io.objectbox.b.a(BoxStoreBuilder.java:147)
       at com.mm.main.app.application.MyApplication.onCreate(MyApplication.java:77)
       at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1017)
       at android.app.ActivityThread.handleBindApplication(ActivityThread.java:4864)
       at android.app.ActivityThread.access$1500(ActivityThread.java:178)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1531)
       at android.os.Handler.dispatchMessage(Handler.java:111)
       at android.os.Looper.loop(Looper.java:194)
       at android.app.ActivityThread.main(ActivityThread.java:5643)
       at java.lang.reflect.Method.invoke(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:372)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:960)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:755)

**Comment by greenrobot** [comment]
This is ObjectBox 0.9.15? So, this is right after app start? Any preceding logs before that error?

**Comment by andrewkkchan** [comment]
This is ObjectBox 0.9.15? : YES
this is right after app start? : YES
Any preceding logs before that error? : NO
Does this occur reproducible/a lot/once only? : A LOT
How many entities do you have?: Around 20 apart from Track.java

**Comment by greenrobot** [comment]
@andrewkkchan I was trying hard to reproduce. I created a test projects with 26 entities, most with a relation (ToOne only admittedly), multi threading, etc.

The closest I could get to was with a high number of threads (200 in my case) with all of them doing db interactions.  In that scenario, I got `io.objectbox.exception.DbMaxReadersExceededException: Could not begin transaction (maximum of read transactions reached) (error code -30790)`.

In that case, I could fix it by increasing the number of maximum readers:

    MyObjectBox.builder().maxReaders(300)

To verify something similar is going on in your code, I would suggest the following:

1. Check the number of active threads while the app is running.
2. Do an occasional `System.out.println(store.diagnose());` and check if the logged entries for readers (e.g. `readers=201`). Are you getting close to 100? Does is correlate with your thread count?
3. Try to increase the reader maximum (see code above)

**Comment by greenrobot** [comment]
No response & cannot reproduce.

**Comment by Yazon2006** [comment]
I faced such kind of problem in my project when I tried to boxStore.runInTx {} some huge amount of transactions and it fails because of some reason (parsing json, npe, etc.). I suppose in my case it was my fault 'cause some transactions were not over and that's why I was not able to run new ones.
Maybe increasing maxReaders count could help like a workaround but as for me the best way is to found a source of such kind of problem and solve it.
Thanks for advices.
